R/unpipe.R

```diff
@@ -43,8 +43,11 @@ is_dot <- function(name) {
 }
 
 unpipe_all <- function(code_expr) {
-  if (length(code_expr) == 1) return(code_expr)
-  if (length(code_expr) == 2 && is.null(code_expr[[2]])) return(code_expr)
+  code_expr_len <- length(code_expr)
+  if (code_expr_len == 0) return(code_expr)
+  if (code_expr_len == 1) return(code_expr)
+  if (code_expr_len == 2 && is.null(code_expr[[2]])) return(code_expr)
+  if (length(code_expr) == 4 && code_expr[[1]] == "function") return(code_expr)
```

Comment:
We still need to recurse on the arguments and function body

```suggestion
  if (code_expr_len == 4 && identical(code_expr[[1]], as.symbol("function"))) {
    # `code_expr` positions: 
    # * 1: as.symbol("function")
    # * 2: list of arguments
    # * 3: body
    # * 4: srcref
    # Remove source ref information to be safe
    code_expr <- code_expr[-4]
  }
```


Current:
```r
code_feedback("function(x, y = a1 %>% b) x + y", "function(x, y = b(a)) x + y")
#> In function(x, y = a1 %>% b) x + y, I expected arguments x, y = b(a) where you wrote arguments x, y = a1 %>% b.

code_feedback("function(x, y) y2 %>% x", "function(x, y) y %>% x")
#> In y2 %>% x, I expected y where you wrote y2.
```

With change:
```r
code_feedback("function(x, y = a1 %>% b) x + y", "function(x, y = b(a)) x + y")
#> In b(a1), I expected a where you wrote a1.

code_feedback("function(x, y) y2 %>% x", "function(x, y) y %>% x")
#> In x(y2), I expected y where you wrote y2.
```

Reply:
What's the best way to detect an empty pairlist value? In other words, to differentiate between `x` and `y` in `as.pairlist(alist(x = , y = 1))`?

With the above changes I still see some odd feedback for empty function arguments:

```r
unpipe_all(quote(function(x, y) x + y))
#> function(, y) x + y
code_feedback("function(z, x, y) x + y", "function(x, y) x + y")
#> In function(z, x, y) x + y, I did not expect your call to () to include x = . You may have included an unnecessary argument, or you may have left out or misspelled an important argument name.
```

Reply:
Nevermind, I think it's that we take a pairlist and cast to a call after recursively unpiping all. 

Instead of

```r
code_expr <- as.call(purrr::map(as.list(code_expr), unpipe_all))
```

we could make sure the pairlist returns as a pairlist

```r
re_call <- if (is.pairlist(code_expr)) as.pairlist else as.call
code_expr <- re_call(purrr::map(as.list(code_expr), unpipe_all))
```

Reply:
With the above change, I'm seeing

```r
code_feedback("function(x, y = a1 %>% b) x + y", "function(x, y = b(a)) x + y")
#> In function(x, y = a1 %>% b) x + y, I expected arguments x, y = b(a) where you wrote arguments x, y = b(a1).
```

which is close (enough?)

Reply:
That works for me

Reply:
The left side function definition will get really big quickly... is there a way to trim it down?

Ex: `In function(x, y = a1 %>% b) x + y, I` to `In y = a1 %>% b, I`?

Reply:
Is it reasonable to assume that if we end up in `wrong_value()` [here](https://github.com/rstudio-education/gradethis/blob/a9c2bf698f538b78ab851184ccf264466f92aa28/R/message_generators.R#L331) with an enclosing call that is a function definition (or more strickly also with a pairlist for `this` and `that`?), then the mistake lies in the argument definitions?